test/control-plane/services/runner.go

```diff
@@ -0,0 +1,463 @@
+			time.Sleep(100 * time.Millisecond)
+			err = step.Validate(lbmap)
+		}
+		if err != nil {
+			t.Fatalf("Test failed: %s", err)
+		}
+	}
+}
+
+// NewGoldenTest creates a test suite from YAML files defining
+// the input states and expected LBMap output.
+//
+// The input states are specified in files state<N>.yaml.
+// These are expected to be a k8s "List", e.g. the format of "kubectl get <res> -o yaml".
+// They're fed to the control-plane in lexicographical order and after each the
+// lbmap state is validated against the expected state as described by lbmap<N>.golden.
+func NewGoldenTest(t *testing.T, nodeName string) (testCase *ServicesTestCase) {
+	testCase = &ServicesTestCase{}
+	testCase.NodeName = nodeName
+
+	if _, err := os.Stat("init.yaml"); err == nil {
+		bs, err := os.ReadFile("init.yaml")
+		if err != nil {
+			t.Fatal(err)
+		}
+		testCase.InitialObjects = unmarshalList(bs)
+	}
+
+	// Construct the test case by parsing all input event files
+	// (<dir>/state*.yaml)
+	ents, err := os.ReadDir(".")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Get the current working directory to construct full path to input files
+	// as cilium-agent will change to the run directory when it starts.
+	workdir, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Sort the entries alphabetically to process the steps in the expected
+	// order.
+	sort.Slice(ents, func(i, j int) bool {
+		return ents[i].Name() < ents[j].Name()
+	})
+
+	for _, ent := range ents {
+		if !strings.HasPrefix(ent.Name(), "state") || !strings.HasSuffix(ent.Name(), ".yaml") {
+			continue
+		}
+
+		stateFile := path.Join(workdir, ent.Name())
+		bs, err := os.ReadFile(stateFile)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		objs := unmarshalList(bs)
+
+		validator := newGoldenLBMapValidator(stateFile, *flagUpdate)
+		testCase.Steps = append(testCase.Steps,
+			NewStep(path.Base(ent.Name()),
+				validator.validate,
+				objs...))
+	}
+	return
+}
+
+type dummyEpSyncher struct{}
+
+func (epSync *dummyEpSyncher) RunK8sCiliumEndpointSync(e *endpoint.Endpoint, conf endpoint.EndpointStatusConfiguration) {
+}
+
+func (epSync *dummyEpSyncher) DeleteK8sCiliumEndpointSync(e *endpoint.Endpoint) {
+}
+
+func setupTestDirectories() string {
+	tempDir, err := ioutil.TempDir("", "cilium-test-")
+	if err != nil {
+		panic("TempDir() failed.")
+	}
+	option.Config.RunDir = tempDir
+	option.Config.StateDir = tempDir
+	return tempDir
+}
+
+type k8sConfig struct{}
+
+func (k8sConfig) K8sAPIDiscoveryEnabled() bool {
+	return false
+}
+
+func (k8sConfig) K8sLeasesFallbackDiscoveryEnabled() bool {
+	return false
+}
+
+type fakeClients struct {
+	core   *fake.Clientset
+	slim   *fakeSlim.Clientset
+	cilium *fakeCilium.Clientset
+	apiext *fakeApiExt.Clientset
+}
+
+type agentHandle struct {
+	d       *cmd.Daemon
+	tempDir string
+}
+
+func startCiliumAgent(nodeName string, clients fakeClients, modConfig func(*option.DaemonConfig)) (datapath.Datapath, agentHandle) {
```

Comment:
Nit: maybe it should be called `mustStartCiliumAgent` given that we expect it to panic when something goes wrong?